src/providers/ibmcloud/mod.rs

```diff
@@ -0,0 +1,82 @@
+//! Metadata fetcher for the IBMCloud provider.
+//!
+//! IBMCloud supports multiple kind of compute nodes, with different
+//! features and peculiarities.
+
+// TODO(lucab): this allows adding 'classic' and 'vpc-gen1' instances too
+//  via auto-detection, if there is a need for that in the future.
+
+use crate::errors::Result;
+use crate::{network, providers};
+use openssh_keys::PublicKey;
+use std::collections::HashMap;
+
+mod gen2;
+
+/// IBMCloud provider (all supported flavors).
+#[derive(Debug)]
+pub enum IbmCloudProvider {
+    /// VPC Generation 2.
+    Gen2(gen2::G2Provider),
+}
+
+impl IbmCloudProvider {
+    /// Build a new client for IBMCloud.
+    ///
+    /// This internally tries to autodetect the infrastructure type
+    /// and mount the relevant config-drive.
+    pub fn try_autodetect() -> Result<Self> {
+        // Auto-detection order: Gen2 (possibly later: classic, Gen1).
+        slog_scope::trace!("trying to autodetect ibmcloud infrastructure type");
+
+        match gen2::G2Provider::try_new() {
+            Ok(g2) => {
+                slog_scope::info!("found metadata for VPC-Gen2 instance");
+                return Ok(IbmCloudProvider::Gen2(g2));
+            }
+            Err(e) => {
+                slog_scope::debug!("VPC-Gen2 autodetection failed: {}", e);
+            }
+        };
+
+        error_chain::bail!("unable to find ibmcloud metadata");
+    }
+}
+
+impl providers::MetadataProvider for IbmCloudProvider {
```

Comment:
Hmm, do we need to switch and dispatch here? Why not just return the `G2Provider` directly to the caller in the `try_autodetect()`? Or do you expect common code to appear here when other cloud variants are added?

Reply:
The latter. There is already a TODO at the top of this file, but I fear you missed it. Where should I move that to be more visible?

Reply:
The top TODO just details that the code is extensible for other cloud types, but doesn't necessarily imply that there would be a lot of shared business logic between implementations. This is fine too though, just trying to avoid boilerplate if we can!

Reply:
Ah, I see your point only now, sorry.
Yes I think I can change `try_autodetect()` to return a generic boxed `MetadataProvider`.

That would avoid the enum and the matching boilerplate. I don't think we have further logic to share here, as you pointed out above.

I'll try to see if the type-checker is happy with that direction.